### ISSUE TITLE
What is Tagged for?

### ISSUE THREAD

**Issue by zenhack** [ocapn/ocapn#52]
Quoting Mark S. Miller (2023-05-21 18:18:34)

>    A bigger pain point would be the explanatory cost: "An OCapN Tagged is
>    a JS Tagged unless the OCapN tag name is 'undefined', in which case it
>    is a JS undefined." Just no. No, on the same grounds on which Christine
>    rejected the omission of symbols. Again, by itself, this is a non-fatal
>    issue. But it would contribute to death by a thousand cuts.

This seems to fundamentally call in to question the value of including
Tagged at all: my understanding has been that the entire point is to
provide a way to extend the data types so that not everything anyone
might want needs to be either included in the raw data. If using them to
translate special js-native types isn't acceptable, what are they for?

Fwiw, I feel like it is fair to also push a bit harder on why using
Tagged to encode symbols is unacceptable, and given the above point of
the confusion, maybe also actually fruitful: I didn't do so at the time
because I don't actually have strong opinions about symbols, but is
there actually a technical reason why it would be problematic to encode
them as Taggeds with a label of "symbol" and a value of that symbol
encoded as a string?

I think hashing out what `Tagged` is for and what it can & can't do for
us is probably actually more important than than either of these
immediate questions of inclusion.

_Originally posted by @zenhack in https://github.com/ocapn/ocapn/issues/50#issuecomment-1556502173_

**Comment by erights** [comment]
Great question! Thinking about it revealed to me that our view of Tagged may indeed be much more limited than other's. As you suspected, it has to do with our layering. Will explain at meeting tomorrow. Thanks!

**Comment by zenhack** [comment]
One thing the other day's presentation (for posterity: https://www.youtube.com/watch?v=htS-2gvY3Cs) didn't clear up for me: the layering issues do seem like they'd make things a bit awkward, but why do those concerns *not* apply to `Set` and `Map`, which we seem to have agreed will be encoded as `Tagged`s?

**Comment by erights** [comment]
Because the `@endo/pass-style` of level just maps them to JS `Tagged` objects with no further interpretation. Only layers build on `@endo/patterns` can see the further classification of some of these as sets, bags, or maps, as classified by `kindOf`. The `@endo/pass-style` of level cannot similarly postpone to higher layers the recognition for something that needs to map to a JS `undefined`. Likewise JS symbols.

Because of peculiarities of the tooling around JS, Agoric is actually operating under bizarre additional constraints that we hope are temporary: https://github.com/endojs/endo/issues/1583 . We're trying not to let these additional temporary constraints leak into OCapN. But it is important if you want to understand the constraints we (Agoric) currently need to operate under, and do.

**Comment by zenhack** [comment]
I guess what I want to understand is, is this just a case of "oops, we already put `undefined` in the wrong layer if we wanted to express it with Tagged, wish we'd done this differently but too late now" or is there a deeper reason why the idea of using Tagged for `undefined` would have problematic even if Agoric didn't already have a bunch of code that does things differently? If the former I'm willing to just say "that's unfortunate, but moving on," but if it's the latter I want to understand why Tagged is fundamentally inadequate for what I understood to be its intended purpose, and what are the actual bounds of its usefulness.

I'd also like to hear from the spritely folks regarding why expressing symbols in terms of Tagged is problematic; while I don't actually care about including it or not, "this datatype is important to our environment" doesn't seem like it should justify needing to bake it into the lowest layer, vs. just encoding symbol foo as `makeTagged('symbol', 'foo')`; it seems like Tagged ought to be able to handle that. (@cwebber, @tsyesika)

**Comment by erights** [comment]
> is there a deeper reason why the idea of using Tagged for undefined would have problematic even if Agoric didn't already have a bunch of code that does things differently

yes. Given only the desire to leave the interpretation of Tagged to higher JS layers that can co-exist and be independent, while the shared JS lower layer must decide what corresponds to a JS `undefined`.

**Comment by zenhack** [comment]
> Given only the desire to leave the interpretation of Tagged to higher JS layers that can co-exist and be independent, while the shared JS lower layer must decide what corresponds to a JS undefined.

My brain is failing to parse this sentence... could you rephrase?

**Comment by erights** [comment]
Given that we desire two layers
   * A lower layer that defines the data model that we all must agree on, for which Tagged is uninterpreted at this layer. More explicitly, the OCapN Tagged is uninterpreted at this layer and translated into a JS Tagged object which is uninterpreted at this layer. All passable JS values *other than* Tagged objects must be translated to OCapN values other than OCapN Tagged. This lower layer is what "we all must agree on" in several senses: between vats, between sites, between languages, and between higher layer modules within a vat.
   * Multiple co-existing independent higher layers, that can be ignorant of each other, each of which defines a set of higher level data types that in encodes into Tagged / recognizes in an encoding of Tagged. We do not need global agreement on these to get global interoperation. The distinction between these higher layers is in terms of what encodings into Taggeds they define, and what they define them to mean. Our `@endo/patters` is an example. Say we also had an independent `@endo/foo`. A foo in one language can be mutually understood by a foo in another language, so the co-existence of foo vs patterns at one site must be able to be reflected in the co-existence of foo and patterns at another site.

**Comment by zenhack** [comment]
I guess what I've been getting at is that it isn't obvious to me why `undefined` couldn't have just included in the second of those layers, and omitted from the first. But it sounds like perhaps the issue is that Agoric has more than one package at the second layer, which don't share a "common" set of recognized tags?

**Comment by erights** [comment]
Agoric does not actually have multiple such second layers. But by the nature of the layer separation, everyone should be able to hav multiple such independent second layers so the parts of system X that know about higher layer B can talk to, and be mutually understood by, the parts of system Y that know about higher layer B. That is inherent in the notion that the lower layer provides Tagged ***as an extension point*** that others can use to extend into higher layers.

Since the JS undefined is not an uninterpreted JS Tagged object since it is distinct from any JS object. Thus, the interpretation of an OCapN Tagged as a JS undefined cannot be postponed to some higher layer that provides *an* (not *the*) interpretation of Taggeds.

Alternatively, if we assumed we must get agreement on only one interpretation of Tagged that must be globally agreed on, then Tagged is not what I would consider an extension point. The multiple co-existing interpretations cannot be, for example, per language, because then a subsystem in language X that understands extension B could not be mutually understood by a subsystem in language Y that also understands extension B. The purpose of such an extension point is not to enable multiple languages to play. It must be an orthogonal dimension of variation.

**Comment by kriskowal** [comment]
Per https://github.com/ocapn/ocapn/wiki/Abstract-Syntax, Agoric and Spritely agree that a single-value tagged data type belongs in the abstract data model.

I move that we close this issue. @tsyesika I believe that if you and I as editors agree, we can close this issue summarily since consensus is captured in https://github.com/ocapn/ocapn/issues/5
